File: go/taking-maximum-energy-from-the-mystic-dungeon.go
func maximumEnergy(energy []int, k int) int {
	var maxEnergy int = energy[len(energy) - 1]

	for lastMagician := len(energy) - k; lastMagician < len(energy); lastMagician++ {
		var energyConsumed int = 0
		for m := lastMagician; m >= 0; m -= k {
			energyConsumed += energy[m]
			maxEnergy = max(maxEnergy, energyConsumed)
		}
	}
	return maxEnergy
}

File: python3/taking-maximum-energy-from-the-mystic-dungeon.py
import math


class Solution:
    def maximumEnergy(self, energy: list[int], k: int) -> int:
        max_energy = -math.inf
        for last_magician in range(len(energy) - k, len(energy)):
            energy_consumed = 0
            for i in range(last_magician, -1, -k):
                energy_consumed += energy[i]
                max_energy = max(energy_consumed, max_energy)
        return max_energy
